core/src/main/java/io/parsingdata/metal/expression/value/Fold.java

```diff
@@ -58,23 +61,24 @@ public Fold(final ValueExpression values, final BinaryOperator<ValueExpression>
         final ImmutableList<Optional<Value>> initial = this.initial != null ? this.initial.eval(graph, encoding) : new ImmutableList<>();
         if (initial.size > 1) { return new ImmutableList<>(); }
         final ImmutableList<Optional<Value>> values = prepareValues(this.values.eval(graph, encoding));
-        if (values.isEmpty() || containsEmpty(values)) { return initial; }
+        if (values.isEmpty() || containsEmpty(values).computeResult()) { return initial; }
         if (!initial.isEmpty()) {
-            return ImmutableList.create(fold(graph, encoding, reducer, initial.head, values));
+            return ImmutableList.create(fold(graph, encoding, reducer, initial.head, values).computeResult());
         }
-        return ImmutableList.create(fold(graph, encoding, reducer, values.head, values.tail));
+        return ImmutableList.create(fold(graph, encoding, reducer, values.head, values.tail).computeResult());
```

Comment:
Not possible, because the passed in pair are in different cases:

1. `initial.head` + `values`
2. `values.head` + `values.tail`
3. `reducedValue.head` + `tail.tail`